tests/rptest/tests/pandaproxy_test.py

```diff
@@ -2129,3 +2132,130 @@ def test_upgrade_and_enable_basic_auth(self, base_release: tuple[int, int],
+            assert len(produce_result["offsets"]) == 1
+            offset = produce_result["offsets"][0]
+            assert "error_code" not in offset, offset["error_code"]
+            if offset["partition"] not in self.partition_offsets:
+                self.partition_offsets[offset["partition"]] = {}
+            self.partition_offsets[offset["partition"]][
+                offset["offset"]] = msg_ascii
+
+            return False
+
+        try:
+            wait_until(do_produce, timeout_sec=self.timeout_sec, backoff_sec=1)
+        except ducktape.errors.TimeoutError as ex:
+            self.done = True
+        except Exception as ex:
+            self.exception = ex
+
+
+class VerifiablePandaproxyConsumer(threading.Thread):
+    def __init__(self,
+                 topic: TopicSpec,
+                 consumer: Consumer,
+                 verifiable_producers: list[VerifiablePandaproxyProducer],
+                 timeout_sec: int,
+                 logger: TestLoggerMaker,
+                 auth_tuple: Optional[tuple[str, str]] = None):
+        threading.Thread.__init__(self)
+        self.daemon = True
+        self.topic = topic
+        self.consumer = consumer
+        self.verifiable_producers = verifiable_producers
+        self.timeout_sec = timeout_sec
+        self.logger = logger
+        self.auth_tuple = auth_tuple
+        self.result_raw = None
+        self.done = False
+        self.exception = None
+
+    def _filter_producer_by_partition_offset(self, partition: int,
+                                             offset: int):
+        try:
+            prod = next(
+                filter(
+                    lambda prod: partition in prod.partition_offsets and offset
+                    in prod.partition_offsets[partition],
+                    self.verifiable_producers))
+            return prod
+        except StopIteration:
+            raise RuntimeError(
+                f"Failed to find producer with partition-offset {partition}-{offset}"
+            )
+
+    def _fetch_message_by_partition_offset(self, partition: int, offset: int):
+        verifiable_producer = self._filter_producer_by_partition_offset(
+            partition, offset)
+        return verifiable_producer.partition_offsets[partition][offset]
+
+    def run(self):
+        sc_res = self.consumer.subscribe([self.topic.name],
+                                         auth=self.auth_tuple)
+        assert sc_res.status_code == requests.codes.no_content
+
+        def do_consume():
+            cf_res = self.consumer.fetch(
+                headers=HTTP_CONSUMER_FETCH_JSON_V2_HEADERS,
+                auth=self.auth_tuple)
+            print(f"fetch result {cf_res} {cf_res.text}")
+            self.logger.debug(f"fetch result {cf_res} {cf_res.text}")
+            if cf_res.status_code != requests.codes.ok:
+                print(f"Issue {cf_res} -- {cf_res.text}")
+            assert cf_res.status_code == requests.codes.ok
```

Comment:
This test is currently failing at this assert because somewhere in Redpanda (maybe our kafka client?), we are not handling leadership transfer. Here is an example of the return error:
```json
fetch result <Response [400]> {"error_code":40002,"message":"not_leader_for_partition"}
Issue <Response [400]> -- {"error_code":40002,"message":"not_leader_for_partition"}
```